lib/statistics/statistical_test/chi_squared_test.rb

```diff
@@ -37,6 +37,59 @@ def self.goodness_of_fit(alpha, expected, observed)
           alternative: p_value <= alpha,
           confidence_level: 1 - alpha }
       end
+
+      # The following three functions serve to calculate a test of independence for contingency
+      # tables of the type
+      #
+      #      A   B
+      # X   20  18
+      # Y    7  35
+      #
+      # They have been tested using 2x2 and 3x3 tables.
+      #
+      def self.test_of_independence(alpha, observed)
+        expected = calculate_expected(observed)
+        df = (observed.size - 1) * (observed.transpose.size - 1)
+        chi_score = calculate_chi_score(observed, expected)
+        probability = 1.0 - Statistics::Distribution::ChiSquared.new(df).cumulative_function(chi_score)
+        p_value = 1.0 - probability
+    
+        {
+          chi_score: chi_score,
+          df: df,
+          probability: probability,
+          p_value: p_value,
+          alpha: alpha,
+          #null: alpha < p_value,
+          #alternative: p_value <= alpha,
+          confidence_level: 1 - alpha,
+          expected: expected
+        }
+      end
+      
+      # For a contingency table of observed values, calculate the expected values
+      def self.calculate_expected(observed)
+        row_sums = observed.map { |row| row.sum.to_f }
```

Comment:
since we assume that the code is a matrix, let's also check that `observed` complies with that.

Reply:
see above comment on `chi_statistic` - currently, `observed` is a nested array, not a Matrix

Reply:
let's also change it to `to_r`